Add module stock_reorder_forecast

gfcapalbo opened the pull request:
Allows to predict date stock levels will reach minimum by
analizying sales volume in a period and therefore to trigger RFQ's ahead
of time",

atchuthan committed:
[FIX] moved field potential_qty to stock_available module

katyukha committed:
stock_reserve: port to 9.0 (#189)

OCA-git-bot committed:
[ADD] setup.py

OCA-git-bot committed:
[UPD] addons table in README.md

gfcapalbo committed:
Add module stock_reorder_forecast

flake8

Tests - first commit

fix

review mods

flake8

tests

remove inheritance from stock to trigger tests

data and some tests

adding tests and data

fixes post test in average calculation and isbuy function

add more tests
correcting date check in query
add priority managment in period selection

defaults expressed as config parameters

removing , defaults that didn't allow config defaults to kick in
fixing config defaults and calls
tests

add config parameter data

more tests

bug fixes

more tests
refactoring
bugfixes

linting
bugfixes

covering new branches

more coverage

lint

flake

add testdb support script

FIx, this took 45 minutes to figure out

fixes

fixes

fixes

new features for client

fix

remove testdb script
manage primary products
clean views
batch creation of on PO wil many POLS

improvements

fix

linting
small bugfixing
refactor
removing query extras used for debugging

fix

fix

Fixed bug in query
new and updated tests

flake/lint

more tests

fix

fix

fixes

gfcapalbo committed:
Add module stock_reorder_forecast

flake8

Tests - first commit

fix

review mods

flake8

tests

remove inheritance from stock to trigger tests

data and some tests

adding tests and data

fixes post test in average calculation and isbuy function

add more tests
correcting date check in query
add priority managment in period selection

defaults expressed as config parameters

removing , defaults that didn't allow config defaults to kick in
fixing config defaults and calls
tests

add config parameter data

more tests

bug fixes

more tests
refactoring
bugfixes

linting
bugfixes

covering new branches

more coverage

lint

flake

add testdb support script

FIx, this took 45 minutes to figure out

fixes

fixes

fixes

new features for client

fix

remove testdb script
manage primary products
clean views
batch creation of on PO wil many POLS

improvements

fix

linting
small bugfixing
refactor
removing query extras used for debugging

fix

fix

Fixed bug in query
new and updated tests

flake/lint

more tests

fix

fix

fixes

NL66278 commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+* Giovanni Francesco Capalbo <giovanni@therp.nl> 
+* Holger Brunn <hbrunn@therp.nl>
+* Hans Van Dijk
```
<hvd400@gmail.com>

gfcapalbo replied:
done

NL66278 commented on stock_reorder_forecast/__openerp__.py:
```diff
@@ -0,0 +1,26 @@
+{
+    "name": "Stock reorder forecast",
+    "version": "9.0.1.0.0",
+    "author": "Therp BV",
```
Add OCA to authors in the usual way: "Therp BV,Odoo Community Association (OCA)"

gfcapalbo replied:
done

NL66278 commented on stock_reorder_forecast/__openerp__.py:
```diff
@@ -0,0 +1,26 @@
+    "category": "Stock",
+    "summary": "Allows to predict date stock levels will reach minimum by
+    analizying sales volume in a period and therefore to trigger RFQ's ahead
+    of time",
```
Keep summary on one line or use either triple quotes, or fully quote each line separately.

NL66278 commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,189 @@
+             "purchase is less than these days away"
+    )
+
+    #called by the cron job every day
```
Put comments on methods in the method they belong to.

NL66278 commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,189 @@
+
+    #called by the cron job every day
+
+    def isbuy(self, product):
```
Add doc string

gfcapalbo replied:
done

NL66278 commented:
@gfcapalbo I think you will have to rebase as well due to recent update of 9.0

gfcapalbo commented:
Seems that there is nothing to rebase, i have created this branch off the latest 9.0 (45 minutes ago) and just added the module

git fetch --all
git rebase origin/9.0
Current branch 9.0-stock_reorder_forecast is up to date.

gfcapalbo commented:
rebased and solved some lint problems (unused js file contained syntax error)

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+Stock Reorder Forecast
+======================
+
+Allows to predict date stock levels will reach minimum by
```
+ 'when'

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+This module allows to create RFQ's by checking the product form and 
+examining the ultimate purchase value.
+
+The ultimate purchase value is the date we foerecast this product will not be
```
forecast

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+examining the ultimate purchase value.
+
+The ultimate purchase value is the date we foerecast this product will not be
+available.It is obtained  by calculating the average sales rate of this 
```
extra space

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+            * Category
+
+if the values are not set on supplier info it will default to values on
+partner, if not set on partner will default to category, if not set anywhere 
```
make clear that with partner the supplier is meant, not some other partner

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+if the values are not set on supplier info it will default to values on
+partner, if not set on partner will default to category, if not set anywhere 
+will default to Hardcoded values (period=365 days).  
```
`ir.config_parameters` with a default value seems better to me

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+launch it multiple times a day.
+
+
+VIew products to see all products with an ultimate order date, for that
```
typo

hbrunn commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+
+VIew products to see all products with an ultimate order date, for that
+interface you can generate a RFQ to deisred date.
```
typo

hbrunn commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,188 @@
+             "purchase is less than these days away"
+    )
+
+    def isbuy(self, product):
```
ormcache this function. Or make this a stored field on products recomputed when the routes change

hbrunn commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,188 @@
+            turnover_average = float_round(
+                turnover_average, self._fields['turnover_average'].digits[1])
+            this = self.browse(product_id)
+            # create a method to see if the product is 'BUY'
```
this has happened already

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    # TODO UPDATE PURCHASE PROPOSAL UPDATE RFQ
```
is this still TODO?

gfcapalbo replied:
no

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+
+    # TODO UPDATE PURCHASE PROPOSAL UPDATE RFQ
+    @api.multi
+    def update_proposal(self, upd_ids=None):
```
remove parameter `upd_ids`

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+    @api.model
+    def create(self, vals):
+        purchase = super(PurchaseOrder, self).create(vals=vals)
+        self.update_proposal(upd_ids=[purchase.id])
```
call this on `purchase`

gfcapalbo replied:
update proposal's signature is (self, upd_ids=None) and that allows me to use it in two different contexts , but i guess i could just call it on purchase, right.

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+        rfqs = self.env['purchase.order'].search([
+            ('id', 'in', upd_ids),
+            ('state', '=', 'draft')
+        ])
```
don't trigger a new search, filter `self` for what you need

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+        ])
+        for rfq in rfqs:
+            for pol in rfq.order_line:
+                pol.product_id.write({'ultimate_purchase': False})
```
a chained `mapped` call will save a plethora of sql statemetns here

hbrunn commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+                pol.product_id.write({'ultimate_purchase': False})
+        # set the partner's ultimate_purchase as the soonest
+        # of all the UP's in the products where he is supplier.
+        sql = self.env['res.partner'].sql_update_partner
```
why isn't this a function that does whatever you need to be done on the partner?

hbrunn commented on stock_reorder_forecast/models/purchase_order_line.py:
```diff
@@ -0,0 +1,59 @@
+            purchase_multiple = self.env['product.product'].PURCHASE_MULTIPLE
+            self.env.cr.execute(sql, (stock_period_max, purchase_multiple,
+                                      self.partner_id.id, self.product_id.id),)
+            rows = self.env.cr.dictfetchall()
```
`dictfetchall` is very expensive. Avoid when possible

hbrunn commented on stock_reorder_forecast/models/purchase_order_line.py:
```diff
@@ -0,0 +1,59 @@
+                product = self.product_id
+                stock = product.virtual_available
+                turnover_average = product.turnover_average
+                self.qty = round(
```
always use `float_round` for quantities

hbrunn commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,65 @@
+        could be done also by searching on seller_ids from the product"""
+        self.env.cr.execute(
+            'select name, array_agg(product_tmpl_id) from product_supplierinfo'
+            ' where name in %s group by name', (tuple(self.ids),))
```
protect against empty ids

hbrunn commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,65 @@
+    """
+
+    @api.multi
+    def _product_supplierinfo(self):
```
this function name doesn't conform the oca naming conventions

hbrunn commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,65 @@
+
+    product_ids = fields.Many2many(
+        'product.template',
+        compute=_product_supplierinfo,
```
use quoted function names to simplify inheritance

hbrunn reviewed (changes requested):
you'll have to replace the therp logo by the oca one.

Then you really need tests for this. For everything. Also add demo data, this will make your live for the tests a lot simpler.

gfcapalbo commented:
writing tests.

hbrunn commented on stock_reorder_forecast/__openerp__.py:
```diff
@@ -0,0 +1,27 @@
+        'purchase'
+    ],
+    "demo": [
+        'tests/data.xml',
```
those usually live in a folder `demo/`. And you forgot to git add the file

gfcapalbo commented:
#tests added and passing. several bugfixes, thank you for your reviews.

hbrunn commented:
check travis and code coverage

rousseldenis commented:
@hbrunn As I see, to avoid unnecessary reviewing, is it possible to add tag work in progress ?

gfcapalbo commented:
@rousseldenis adding more tests and rebasing, thank you

gfcapalbo committed:
Merge remote-tracking branch 'origin/9.0' into 9.0-stock_reorder_forecast

gfcapalbo committed:
flake8

gfcapalbo committed:
flake8

gfcapalbo committed:
lint , rem line

gfcapalbo committed:
lint , rem line

gfcapalbo committed:
fix a problem in the query and have the correct testas for calculating
suppliers

gfcapalbo committed:
fix a problem in the query and have the correct testas for calculating
suppliers

gfcapalbo committed:
very basic sequence managment for product supplier, allows user to manually see
and insert sequence value instead of using widget handle, ideally in future
could add a is_primary field and highlight the current primary supplier in form
and perhaps also a button "set as primary" that would rearrange sequences
making the supplier the primary.

gfcapalbo committed:
very basic sequence managment for product supplier, allows user to manually see
and insert sequence value instead of using widget handle, ideally in future
could add a is_primary field and highlight the current primary supplier in form
and perhaps also a button "set as primary" that would rearrange sequences
making the supplier the primary.

gfcapalbo committed:
flake8

gfcapalbo committed:
flake8

gfcapalbo committed:
fixes

gfcapalbo committed:
fixes

gfcapalbo committed:
last tests and mods ,. quantities fixed

gfcapalbo committed:
last tests and mods ,. quantities fixed

gfcapalbo committed:
flake8

gfcapalbo committed:
flake8

gfcapalbo committed:
last important fixes

gfcapalbo committed:
last important fixes

gfcapalbo commented:
@rousseldenis  if i ever implemented that it would be a good idea. for now it just removes the handle widget and shows sequence in the view. Hardly a  feature just a view improvement.

rousseldenis commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+The turnover average and the ultimate purchase derived by it are calculated in
+bulk by a daily cron job.
+
+==========
```
Can you remove this ? (OCA maintainer-tools template)

rousseldenis commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+==========
+Usage
+==========
```
Can you lower = cases ?

rousseldenis commented on stock_reorder_forecast/README.rst:
```diff
@@ -0,0 +1,87 @@
+You can also view from the partner/supplier form all products ordered by this
+partner.
+
+
```
Missing bug tracker (https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst)

rousseldenis commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,222 @@
+             "purchase is less than these days away"
+    )
+    is_buy = fields.Boolean("Does the product have Buy Route?",
+                            compute='_compute_isbuy', stored=True)
```
This should be 'store' and not 'stored'

gfcapalbo replied:
True, thank you. Unknown attributes are ignored by the orm so effectivley this is not-stored (as default)

rousseldenis commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,222 @@
+            stock_days = int(float_round((((
+                this.virtual_available or 0
+                ) - stock_period_min) / turnover_average) + .5, 0))
+            if self.has_purchase_draft():
```
Is 'self' correct here ? Should be this.

gfcapalbo replied:
fixed

rousseldenis commented on stock_reorder_forecast/data/ir_config_parameter.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<openerp>
+    <data noupdate="1">
+        <record id="config_turnover_period" model="ir.config_parameter">
```
I am not sure using ir.config_parameter is a good approach (for these parameters) as it is not multi company compatible.

gfcapalbo replied:
These are last resort values. the shouldn't be used normally. Every company should have it's own values. therefore having the same generic for all companies makes sense.

rousseldenis commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,222 @@
+
+    def has_purchase_draft(self):
+        sql = """
+            select count(*) from purchase_order_line
```
If using count, prefer count(id) as it is more efficient.

gfcapalbo replied:
fixed

rousseldenis commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,222 @@
+            select count(*) from purchase_order_line
+                where product_id = %s
+                and order_id in (select id from purchase_order where
+                state='draft')
```
You are never sure of the amount of draft purchase orders, so I'd rather write  query like this :

`SELECT COUNT(pol.id) FROM purchase_order_line pol 
JOIN purchase_order po ON pol.order_id = po.id
WHERE po.state = 'draft'
AND pol.product_id = %s`

rousseldenis commented on stock_reorder_forecast/models/product_product.py:
```diff
@@ -0,0 +1,222 @@
+            [('key', '=', 'default_turnover_period')])[0].value)
+
+    @api.model
+    def calc_purchase_date(self):
```
As I said here over, this cannot work in multi company environment (even more in a query) !

gfcapalbo replied:
looking into changing the query to account multicompany.

gfcapalbo replied:
multicompany adaptation of the query is not implemented. added to readme.
Currently the cron job gives us global stats, that may be useful for some multicompany configurations , but not all.

https://github.com/OCA/stock-logistics-warehouse/pull/222/commits/d9f90d95492c790e4851748eab4946f3b5ddb19a

rousseldenis commented on stock_reorder_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,33 @@
+    def write(self, vals):
+        result = super(PurchaseOrder, self).write(vals)
+        if result:
+            self.update_proposal()
```
Shouldn't be 'result' ?

gfcapalbo replied:
No, write returns "True" not a record.

rousseldenis replied:
Oups, sorry!

rousseldenis commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,98 @@
+'''
+Additional fields for the purchase proposal
+'''
+from openerp import fields, api, models
```
Can you reorder imports ?

gfcapalbo replied:
fixed

rousseldenis commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,98 @@
+    @api.model
+    def sql_update_partner(self):
+        sql = """
+        UPDATE res_partner RP
```
Had you experienced orm performance problems to do this through sql ?

gfcapalbo replied:
this is significantly more performant via sql.   ORM implementation would had been more readable but the time of the cronjob , especially on large db's would had been quit long. Definetley more performant. on a average DB (1000+ products) this fetches all the information we need in one pass in under 2 minutes.  the ORM version could be above 10 minutes. THe function might also be triggered manually.

rousseldenis commented on stock_reorder_forecast/models/res_partner.py:
```diff
@@ -0,0 +1,98 @@
+            if this.id in partner2products.keys():
+                this.primary_product_ids = partner2products[this.id]
+
+    def _compute_product_supplierinfo(self):
```
Should be decorated @api.multi

gfcapalbo replied:
fixed

rousseldenis reviewed (changes requested):
Code review, not tested on runbot
Great job !

rousseldenis commented on stock_reorder_forecast/views/partner_view.xml:
```diff
@@ -0,0 +1,72 @@
+            </field>
+        </record>
+
+        <record id="suplier_show_sequence_tree" model="ir.ui.view">
```
@gfcapalbo IMO this should not be part of this module as it changes base behaviour and get out of the scope (waiting other opinions ;-) )

gfcapalbo replied:
I see it as a small cosmetic change, not a functional change.
We are just changing the visualization and input of sequence on one view to improve sequence visualization and modification.  the separate module would be interesting  by changing all handle widgets  on all models  by overwriting fields_view_get.
I vote it stays.  other opinions welcome.   @hbrunn  @NL66278 ?

gfcapalbo committed:
Fix Readme OCA, style

review fixes

gfcapalbo committed:
Fix Readme OCA, style

review fixes

gfcapalbo committed:
Multicompany TODO

gfcapalbo committed:
Multicompany TODO

gfcapalbo committed:
More Readme info

gfcapalbo committed:
More Readme info

gfcapalbo committed:
Removing "country id" extension, slipped in this module, but was unrelated

gfcapalbo committed:
Removing "country id" extension, slipped in this module, but was unrelated

gfcapalbo committed:
Stock_period_min reestablished on product

gfcapalbo committed:
Stock_period_min reestablished on product

gfcapalbo committed:
fix create dict in test
remove duplicate definition

gfcapalbo committed:
fix create dict in test
remove duplicate definition

hbrunn commented:
this breaks because this wants to be the comodel parameter: https://github.com/OCA/stock-logistics-warehouse/pull/222/files#diff-665bec2e1099a6cd5883316e011db8daR101
this causes a warning: https://github.com/OCA/stock-logistics-warehouse/pull/222/files#diff-ffadef523028be6685cb04cd29d57202R52

gfcapalbo committed:
fix comodel name and add comma

gfcapalbo committed:
fix comodel name and add comma

gfcapalbo committed:
remove extra entry in create dict

gfcapalbo committed:
remove extra entry in create dict

gfcapalbo committed:
resolving issue with wrong purchase lines created, negative quantities where
ordered anyways. now set to 0

gfcapalbo committed:
resolving issue with wrong purchase lines created, negative quantities where
ordered anyways. now set to 0

gfcapalbo committed:
Adding test to account case of qty <0 or =0

gfcapalbo committed:
Adding test to account case of qty <0 or =0

gfcapalbo committed:
linting

gfcapalbo committed:
linting

gfcapalbo committed:
[FIX] no more empty purchase orders

gfcapalbo committed:
Merge branch '9.0-stock_reorder_forecast' of https://github.com/gfcapalbo/stock-logistics-warehouse into 9.0-stock_reorder_forecast

gfcapalbo committed:
stock_period_max connected to product.

gfcapalbo committed:
Bypass automatic procurement order generation

gfcapalbo committed:
fix

gfcapalbo committed:
consistent return with no po created

hbrunn commented on stock_reorder_forecast/models/procurement_order.py:
```diff
@@ -0,0 +1,14 @@
+
+    @api.model
+    def _run(self, procurement):
+        return []
```
this is bad. Very bad, because it turns off procurements entirely for everything. That's not what you want

gfcapalbo committed:
Non destructive procurement inheritance

hbrunn commented on stock_reorder_forecast/models/procurement_order.py:
```diff
@@ -0,0 +1,30 @@
+    def _run(self, procurement):
+        if procurement.rule_id and procurement.rule_id.action == 'buy':
+            return  super(ProcurementOrder, self.with_context(
+                skip_po=True))._run(procurement)
```
`self` is just a variable name, so if you say here `self = self.with_context(...)`, you can just fall through and have no code duplication

hbrunn commented on stock_reorder_forecast/models/procurement_order.py:
```diff
@@ -0,0 +1,30 @@
+    @api.multi
+    def make_po(self):
+        for procurement in self:
+            if self.context['skip_po'] == True:
```
this breaks if the key doesn't exist, and use truthyness

hbrunn commented on stock_reorder_forecast/models/procurement_order.py:
```diff
@@ -0,0 +1,30 @@
+    def make_po(self):
+        for procurement in self:
+            if self.context['skip_po'] == True:
+                return []
```
read https://github.com/OCA/OCB/blob/9.0/addons/procurement/procurement.py#L195 and think again if this is a good idea

gfcapalbo committed:
a better solution using autorun defer

gfcapalbo committed:
easier solution

gfcapalbo committed:
final, testing

gfcapalbo committed:
extra check for SO

gfcapalbo committed:
better condition

gfcapalbo committed:
Last fix, simplified

hbrunn committed:
[IMP] some view improvements; show turnover of variants on templates

hbrunn committed:
[FIX] flake8

hbrunn committed:
[ADD] more view fiddling

hbrunn committed:
[FIX] logic error concerning use of max vs greatest

[DEL] redundant subqueries slowing down the query a lot

gfcapalbo committed:
Merge pull request #1 from hbrunn/9.0-stock_reorder_forecast

9.0 stock reorder forecast

gfcapalbo committed:
better wizard look

gfcapalbo committed:
Do not allow the generation of empty POs

gfcapalbo committed:
fix for issue 3194

gfcapalbo committed:
fix issue 3383

gfcapalbo committed:
[FIX] now orders all products
[FIX] cleanup view

gfcapalbo committed:
[IMP] Query now returns product_product not product_template

gfcapalbo committed:
Merge remote-tracking branch 'oca/9.0' into 9.0-stock_reorder_forecast
